### PR TITLE
Guardian Light Backend Support

### PR DESCRIPTION
**Pull request by tjmw**
## What are you doing in this PR?

Following on from #6498 this PR adds support for purchasing Guardian Light on the backend.

**Note: does not include the thank you page or completed thank you/failure emails, this work will be done in future PRs.**

[**Trello Card**](https://trello.com/c/V7r1ALrC/1211-checkout-support-for-new-product-backend)

## Why are you doing this?

So that users can purchase Guardian Light and the correct data is added to Zuora.

## How to test

I've been able to put through a transaction successfully in CODE:

- [x] The support-workers state machine execution completes successfully
- [x] The correct cookie is dropped
- [x] Thank you email message sent to the membership-workflow queue
- [x] Data added to the `SupporterProductData` Dynamo table

## Accessibility test checklist

-  [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-  [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-  [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-  [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

## Screenshots

Screenshot 2024-11-21 at 13 16 27

Screenshot 2024-11-20 at 16 49 58

### CODE REVIEW

File: support-models/src/main/scala/com/gu/support/workers/Products.scala [hunk 1]
```diff
@@ -81,6 +81,14 @@ case class GuardianWeekly(
   override def describe: String = s"$billingPeriod-GuardianWeekly-$fulfilmentOptions-$currency"
 }
 
+case class GuardianLight(
+    amount: BigDecimal,
```
Comment (tjmw) [inline support-models/src/main/scala/com/gu/support/workers/Products.scala:85]: Maybe not needed?

File: support-config/src/main/resources/touchpoint.CODE.conf [hunk 2]
```diff
@@ -60,6 +60,10 @@ touchpoint.backend.environments {
             annual.contribution.productRatePlanChargeId = "8ad096ca858682bb0185881568385d73"
         }
       }
+      guardianlight {
```
Comment (tjmw) [inline support-config/src/main/resources/touchpoint.CODE.conf:63]: TODO: add this config for PROD too

File: support-frontend/app/controllers/CreateSubscriptionController.scala [hunk 3]
```diff
@@ -379,6 +379,11 @@ class CreateSubscriptionController(
         )
       case _: Paper => List.empty
       case _: GuardianWeekly => List.empty
+      // TODO: Confirm this
```
Comment (tjmw) [inline support-frontend/app/controllers/CreateSubscriptionController.scala:382]: Remove TODO, this is confirmed.

File: support-models/src/main/scala/com/gu/support/catalog/Product.scala [hunk 4]
```diff
@@ -109,6 +109,31 @@ case object SupporterPlus extends Product {
     )
 }
 
+case object GuardianLight extends Product {
+  private def productRatePlan(
+      id: String,
+      billingPeriod: BillingPeriod,
+      fulfilmentOptions: FulfilmentOptions = NoFulfilmentOptions,
```
Comment (tjmw) [inline support-models/src/main/scala/com/gu/support/catalog/Product.scala:116]: Remove both of these and hardcode in the function body.

File: support-models/src/main/scala/com/gu/support/workers/ProductTypeRatePlans.scala [hunk 5]
```diff
@@ -39,6 +39,14 @@ object ProductTypeRatePlans {
       .getOrElse(environment, Nil)
       .find(productRatePlan => productRatePlan.billingPeriod == product.billingPeriod)
 
+  def guardianLightRatePlan(
+      product: GuardianLight,
+      environment: TouchPointEnvironment,
+  ): Option[ProductRatePlan[catalog.GuardianLight.type]] =
+    catalog.GuardianLight.ratePlans
+      .getOrElse(environment, Nil)
+      .find(productRatePlan => productRatePlan.billingPeriod == product.billingPeriod)
```
Comment (tjmw) [inline support-models/src/main/scala/com/gu/support/workers/ProductTypeRatePlans.scala:48]: Revisit this now we've removed `billingPeriod` from the product. Maybe just grab the first (assuming there will only be one).

File: support-models/src/main/scala/com/gu/support/workers/Products.scala [hunk 6]
```diff
@@ -81,6 +81,13 @@ case class GuardianWeekly(
   override def describe: String = s"$billingPeriod-GuardianWeekly-$fulfilmentOptions-$currency"
 }
 
+case class GuardianLight(
+    currency: Currency,
+    billingPeriod: BillingPeriod,
```
Comment (tjmw) [inline support-models/src/main/scala/com/gu/support/workers/Products.scala:86]: Remove this.

File: support-models/src/main/scala/com/gu/support/workers/states/CreateZuoraSubscriptionState.scala [hunk 7]
```diff
@@ -60,6 +60,13 @@ object CreateZuoraSubscriptionProductState {
       salesForceContact: SalesforceContactRecord,
   ) extends CreateZuoraSubscriptionProductState
 
+  case class GuardianLightState(
+      user: User, // Is this user property needed?
```
Comment (tjmw) [inline support-models/src/main/scala/com/gu/support/workers/states/CreateZuoraSubscriptionState.scala:64]: Nope, not needed.

File: support-workers/src/main/scala/com/gu/acquisitions/AcquisitionDataRowBuilder.scala [hunk 8]
```diff
@@ -257,6 +259,15 @@ object AcquisitionDataRowBuilder {
...
           None,
         )
+      case s: SendThankYouEmailGuardianLightState =>
+        AcquisitionTypeDetails(
+          paymentMethod = Some(s.paymentMethod),
+          promoCode = None,
+          readerType = Direct,
+          acquisitionType = Purchase,
+          zuoraAccountNumber = Some(s.accountNumber),
+          zuoraSubscriptionNumber = Some(s.subscriptionNumber),
```
Comment (tjmw) [inline support-workers/src/main/scala/com/gu/acquisitions/AcquisitionDataRowBuilder.scala:269]: Maybe check the defaults here.

File: support-workers/src/main/scala/com/gu/emailservices/FailedContributionEmailFields.scala [hunk 9]
```diff
@@ -20,6 +20,9 @@ object FailedEmailFields {
   def paper(email: String, identityUserId: IdentityUserId): EmailFields =
     failedEmailFields("paper-failed", email, identityUserId)
 
+  def guardianLight(email: String, identityUserId: IdentityUserId): EmailFields =
+    failedEmailFields("guardian-light-failed", email, identityUserId)
```
Comment (tjmw) [inline support-workers/src/main/scala/com/gu/emailservices/FailedContributionEmailFields.scala:24]: TODO: figure out what needs adding to membership-workflow.

Comment (tjmw) [reply support-workers/src/main/scala/com/gu/emailservices/FailedContributionEmailFields.scala:24]: I've added a placeholder PR. We don't have the campaign IDs yet.

**Review by rupertbates — APPROVED** [review]
Great work getting through all this 👍

**Pull request merged**